beacon-chain/sync/rpc_blob_sidecars_by_root.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/p2p/types"
+	"github.com/prysmaticlabs/prysm/v4/cmd/beacon-chain/flags"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/v4/monitoring/tracing"
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	"go.opencensus.io/trace"
+)
+
+func blobMinReqEpoch(finalized, current primitives.Epoch) primitives.Epoch {
+	// max(finalized_epoch, current_epoch - MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUESTS, DENEB_FORK_EPOCH)
+	denebFork := params.BeaconConfig().DenebForkEpoch
+	var reqWindow primitives.Epoch
+	if current > params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest {
+		reqWindow = current - params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	}
+	if finalized >= reqWindow && finalized > denebFork {
+		return finalized
+	}
+	if reqWindow >= finalized && reqWindow > denebFork {
+		return reqWindow
+	}
+	return denebFork
+}
+
+// blobSidecarByRootRPCHandler handles the /eth2/beacon_chain/req/blob_sidecars_by_root/1/ RPC request.
+// spec: https://github.com/ethereum/consensus-specs/blob/a7e45db9ac2b60a33e144444969ad3ac0aae3d4c/specs/deneb/p2p-interface.md#blobsidecarsbyroot-v1
+func (s *Service) blobSidecarByRootRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.blobSidecarByRootRPCHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, ttfbTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRootName[1:]) // slice the leading slash off the name var
+	ref, ok := msg.(*types.BlobSidecarsByRootReq)
+	if !ok {
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+
+	blobIdents := *ref
+	if err := validateBlobByRootRequest(blobIdents); err != nil {
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		return err
+	}
+	// Sort the identifiers so that requests for the same blob root will be adjacent, minimizing db lookups.
+	sort.Sort(blobIdents)
+
+	batchSize := flags.Get().BlobBatchLimit
+	var ticker *time.Ticker
+	if len(blobIdents) > batchSize {
+		ticker = time.NewTicker(time.Second)
+	}
+	minReqEpoch := blobMinReqEpoch(s.cfg.chain.FinalizedCheckpt().Epoch, slots.ToEpoch(s.cfg.clock.CurrentSlot()))
+
+	buff := struct {
+		root [32]byte
+		scs  []*eth.BlobSidecar
+	}{}
+	for i := range blobIdents {
+		if err := ctx.Err(); err != nil {
+			closeStream(stream, log)
+			return err
+		}
+
+		// Throttle request processing to no more than batchSize/sec.
+		if i != 0 && i%batchSize == 0 && ticker != nil {
+			<-ticker.C
+		}
+		s.rateLimiter.add(stream, 1)
```

Comment:
Should you increment the rate limiter before a potentially blocking event? i.e. if the chan is buffered/full then it is blocking _before_ the rate limiter was incremented which could be an issue.

Reply:
good catch, will fix.

Reply:
there's no buffered chan or otherwise non-deterministic block here fwiw, this is just a ticker.